x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+     *
+     * @param token The OpenIdConnectToken to consume
+     * @return a {@link JWTClaimsSet} with the OP claims for the user
+     * @throws ElasticsearchSecurityException if the response is invalid in any way
+     */
+    public JWTClaimsSet authenticate(OpenIdConnectToken token) {
+        try {
+            AuthenticationResponse authenticationResponse = AuthenticationResponseParser.parse(new URI(token.getRedirectUrl()));
+            Nonce expectedNonce = new Nonce(token.getNonce());
+            State expectedState = new State(token.getState());
+            if (logger.isTraceEnabled()) {
+                logger.trace("OpenID Connect Provider redirected user to [{}]. Expected Nonce is [{}] and expected State is [{}]",
+                    token.getRedirectUrl(), expectedNonce, expectedState);
+            }
+            if (authenticationResponse instanceof AuthenticationErrorResponse) {
+                ErrorObject error = ((AuthenticationErrorResponse) authenticationResponse).getErrorObject();
+                throw new ElasticsearchSecurityException("OpenID Connect Provider response indicates authentication failure." +
+                    "Code=[{}], Description=[{}]", error.getCode(), error.getDescription());
+            }
+            final AuthenticationSuccessResponse response = authenticationResponse.toSuccessResponse();
+            validateState(expectedState, response.getState());
+            validateResponseType(response);
+            JWT idToken;
+            AccessToken accessToken;
+            if (rpConfig.getResponseType().impliesCodeFlow()) {
+                final AuthorizationCode code = response.getAuthorizationCode();
+                Tuple<AccessToken, JWT> tokens = exchangeCodeForToken(code);
+                accessToken = tokens.v1();
+                idToken = tokens.v2();
+            } else {
+                idToken = response.getIDToken();
+                accessToken = response.getAccessToken();
+            }
+            final JWTClaimsSet idTokenClaims = validateAndParseIdToken(idToken, expectedNonce);
+            if (opConfig.getUserinfoEndpoint() != null) {
+                validateAccessToken(accessToken, idToken, rpConfig.getResponseType().impliesCodeFlow());
+                final JWTClaimsSet userInfoClaims = getUserInfo(accessToken);
+                final JSONObject combinedClaims = idTokenClaims.toJSONObject();
+                combinedClaims.merge(userInfoClaims.toJSONObject());
+                return (JWTClaimsSet.parse(combinedClaims));
+            } else {
+                return idTokenClaims;
+            }
+
+        } catch (ElasticsearchSecurityException e) {
+            // Don't wrap in a new ElasticsearchSecurityException
+            throw e;
+        } catch (Exception e) {
+            logger.debug("Failed to consume the OpenID connect response", e);
+            throw new ElasticsearchSecurityException("Failed to consume the OpenID connect response");
+        }
```

Comment:
This block can be replaced with `ExceptionsHelper.convertToElastic`.

Reply:
I don't think so since we're using a ElasticsearchSecurityException for the response code of a 403 (IIUC)